pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go

```diff
@@ -0,0 +1,198 @@
+package agentpoller
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/cli"
+)
+
+type AgentPoller struct {
+	Connection cli.Connection
+	VmiUID     types.UID
+	domainData DomainData
+	active     bool
+	trigger    chan *api.Domain
+	pollTime   time.Duration
+	callback   QemuAgentCallback
+}
+
+type DomainData struct {
+	name       string
+	aliases    []string
+	aliasToMac map[string]string
+	interfaces []api.InterfaceStatus
+}
+
+type Result struct {
+	Page       int
+	Interfaces []Interface `json:"return"`
+}
+type Interface struct {
+	MAC  string `json:"hardware-address"`
+	IPs  []IP   `json:"ip-addresses"`
+	Name string `json:"name"`
+}
+
+type IP struct {
+	IP     string `json:"ip-address"`
+	Type   string `json:"ip-address-type"`
+	Prefix int    `json:"prefix"`
+}
+
+type QemuAgentCallback func(string, *[]api.InterfaceStatus)
+
+func CreatePoller(connecton cli.Connection, vmiUID types.UID, callback QemuAgentCallback, qemuAgentPollerInterval *time.Duration) *AgentPoller {
+	p := &AgentPoller{
+		Connection: connecton,
+		VmiUID:     vmiUID,
+		pollTime:   *qemuAgentPollerInterval,
+		callback:   callback,
+	}
+	p.create()
+	return p
+}
+
+func (p *AgentPoller) create() {
+	p.trigger = make(chan *api.Domain, 5)
+
+	go func() {
+		for {
+			log.Log.Info("Qemu agent poller started")
+			select {
+			case domain := <-p.trigger:
+				if !p.active {
+					continue
+				}
+
+				if domain != nil {
+					p.domainData = p.createDomainData(domain)
+				}
+
+				cmdResult, err := p.pollQemuAgent(p.domainData.name)
+				if err != nil {
+					log.Log.Reason(err).Error("Qemu agent poller error")
+					time.Sleep(p.pollTime * time.Second)
+					p.trigger <- nil
+					continue
+				}
+				interfaceStatuses := p.GetInterfaceStatuses(cmdResult)
+
+				if !reflect.DeepEqual(p.domainData.interfaces, interfaceStatuses) {
+					p.domainData.interfaces = interfaceStatuses
+					if p.active {
+						p.callback(p.domainData.name, &interfaceStatuses)
+					}
+				}
+				time.Sleep(time.Duration(p.pollTime) * time.Second)
+				p.trigger <- nil
+			}
+		}
+	}()
+}
+
+func (p *AgentPoller) GetInterfaceStatuses(cmdResult string) []api.InterfaceStatus {
+	parsedResult := parseAgentReply(cmdResult)
+	macToIp, macToIps, macToName := processAgentReply(parsedResult)
+	return p.createInterfaceStatuses(macToIp, macToIps, macToName)
+}
+
+func (p *AgentPoller) UpdateDomain(domain *api.Domain) {
```

Comment:
This can't reliably work. You will have to use a lock. golang makes heavy use of thread-local copies in go-routines. You will see unpredictable results if you don't wrap it in a lock.